### PR TITLE
[CARBONDATA-1940][PreAgg] Fixed bug for creation of preaggregate table with group by clause

### PR DESCRIPTION
**Pull request by geetikagupta16**
1. Refactored code to create pre-aggregate table with group by clause
2. Added related test cases

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2331/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1118/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2560/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateUtil.scala [hunk 1]
```diff
@@ -144,6 +144,16 @@ object PreAggregateUtil {
...
     }
+    groupByExp map {
+      case attr: AttributeReference =>
+        fieldToDataMapFieldMap += getField(attr.name,
+          attr.dataType,
+          parentColumnId = carbonTable.getColumnByName(parentTableName, attr.name).getColumnId,
+          parentTableName = parentTableName,
+          parentDatabaseName = parentDatabaseName, parentTableId = parentTableId)
+      case _ =>
+        throw new MalformedCarbonCommandException(s"Expression in group by clause is incorrect")
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateUtil.scala]: Please update the exception message 
throw new MalformedCarbonCommandException(s"Unsupported Function in select Statement:${
          selectStmt }

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala [hunk 2]
```diff
@@ -1264,13 +1264,16 @@ object CarbonPreAggregateDataLoadingRules extends Rule[LogicalPlan] {
...
     plan transform {
-      case aggregate@Aggregate(_, aExp, _) if validateAggregateExpressions(aExp) =>
+      case aggregate@Aggregate(groupingExpressions, aExp, _) if validateAggregateExpressions(aExp) =>
         aExp.foreach {
           case alias: Alias =>
             validExpressionsMap ++= validateAggregateFunctionAndGetAlias(alias)
           case _: UnresolvedAlias =>
           case namedExpr: NamedExpression => validExpressionsMap.put(namedExpr.name, namedExpr)
         }
+        groupingExpressions foreach {
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: I don't get the purpose of this change data loading rules is only to update the average column as average columns data will be stored in two column(sum, count), can u please explain why we need this change??

Comment (geetikagupta16) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: This change is required for adding grouping expressions in logical plan for data loading in pre aggregate table as now expressions in group clause will also be treated as columns of pre aggregate table

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2345/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2564/

**Comment by geetikagupta16** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2353/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala (hunk 2 above)
Comment (kumarvishal09) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: Ok, This is to handle when column is on grouping expression but not on projection, as we are adding column in data map user can create aggregation data map by adding that column in projection list.  But it is OK to support this type of expression in pre aggregate .

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1136/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala (hunk 2 above)
Comment (geetikagupta16) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateRules.scala]: As per CARBONDATA-1940, it was throwing an exception when grouping columns were not added in projection of preaggregate table

**Comment by BJangir** [comment]
@geetikagupta16 , please handle filter columns also . for example . select country ,sum( case when year=2010 salary else 0 end)  from tbl where year=2010 or year=2011 group by   country.   in this case aggregate datamap should have country,year ,salary columns .

**Comment by kumarvishal09** [comment]
@BJangir Currently we are not supporting filter while creating the aggregation data map. In future we may support. User can add filter column on projection for above use case.

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1327/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2563/

**Comment by geetikagupta16** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2578/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1344/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2786/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1390/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2625/

**Comment by kumarvishal09** [comment]
@geetikagupta16 Please rebase

**Comment by geetikagupta16** [comment]
@kumarvishal09 Rebase is done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1400/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2795/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2635/

**Comment by geetikagupta16** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2651/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1419/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2676/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1442/

**Comment by kunal642** [comment]
@geetikagupta16 can you squash the commits.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1453/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2688/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2825/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2707/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1474/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2840/

**Comment by geetikagupta16** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2715/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1482/

**Comment by geetikagupta16** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2845/

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**